lenskit-core/src/main/java/org/lenskit/data/dao/file/StaticFileDAOProvider.java

```diff
@@ -0,0 +1,191 @@
+ * published by the Free Software Foundation; either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+package org.lenskit.data.dao.file;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ListMultimap;
+import org.lenskit.data.dao.DataAccessException;
+import org.lenskit.data.dao.DataAccessObject;
+import org.lenskit.data.dao.EntityCollectionDAOBuilder;
+import org.lenskit.data.entities.*;
+import org.lenskit.util.io.ObjectStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Provider;
+import java.io.IOException;
+import java.lang.ref.SoftReference;
+import java.net.URI;
+import java.util.*;
+
+/**
+ * Layout and builder for DAOs backed by static files.  This is used to read CSV files
+ * and the like; it controls a composite DAO that reads from files, caches them in
+ * memory, and can compute some derived entities from others (e.g. extracting items
+ * from the item IDs in a rating data set).
+ */
+public class StaticFileDAOProvider implements Provider<DataAccessObject> {
+    private static final Logger logger = LoggerFactory.getLogger(StaticFileDAOProvider.class);
+    private List<EntitySource> sources;
+    private ListMultimap<EntityType, TypedName<?>> indexedAttributes;
+    private transient volatile SoftReference<DataAccessObject> cachedDao;
+
+    /**
+     * Construct a new data layout object.
+     */
+    public StaticFileDAOProvider() {
+        sources = new ArrayList<>();
+        indexedAttributes = ArrayListMultimap.create();
+    }
+
+    /**
+     * Add a collection data source.
+     * @param data The entities to add.
+     */
+    public void addSource(Collection<? extends Entity> data) {
+        sources.add(new CollectionEntitySource(data));
+    }
+
+    private void addSource(EntitySource source) {
+        sources.add(source);
+    }
+
+    /**
+     * Index entities by an attribute.
+     * @param type The entity type to index.
+     * @param attr The attribute to index.
+     */
+    public void addIndex(EntityType type, TypedName<?> attr) {
+        indexedAttributes.put(type, attr);
+    }
+
+    /**
+     * Get the data access object. This method is thread-safe.
+     * @return The access object.
+     */
+    @Override
+    public DataAccessObject get() {
+        SoftReference<DataAccessObject> cache = cachedDao;
+        DataAccessObject dao = cache != null ? cache.get() : null;
```

Comment:
Yes, for now. Right now I am trying to get the public API and a basic, correct implementation down.

I would like to add significantly more optimization, decrease memory usage, etc. once we have that in place.